src/runtime/formal/ops/ops.cc

```diff
@@ -682,36 +556,24 @@ static void take(CVMArgValue x,
 
     if(axis < 0) { axis += xndim; }
 
-    std::vector<size_t> x_shape_size(xndim, 1), indices_shape_size(indices_ndim, 1);
-    for (int i = xndim-2; i >= 0; --i) {
-      x_shape_size[i] = x_shape_size[i+1] * X_shape[i+1];
-    }
-    for (int i = indices_ndim-2; i >= 0; --i) {
-      indices_shape_size[i] = indices_shape_size[i+1] * Indices_shape[i+1];
-    }
-    for (size_t i = 0; i < Y_shape.Size(); ++i) {
-      size_t oi = i, xi = 0, idxi = 0;
-      for(int j = yndim - 1; j>=0; --j){
-        size_t col = oi % Y_shape[j];
-        oi /= Y_shape[j];
-        if (axis <= j && j < axis+indices_ndim) {
-          idxi += col * indices_shape_size[j - axis];
-        } else {
-          int xidx = j < axis ? j : j - indices_ndim + 1;
-          xi += col * x_shape_size[xidx];
-        }
-
-        if (axis == j) {
-          int64_t idxx = std::min(std::max(indices_data[idxi], 0), 
-              (int32_t)X_shape[j]-1);
-          xi += idxx * x_shape_size[j];
-        }
+    for (Indices yIdx(Y_shape), xIdx(X_shape), idxIdx(Indices_shape);
+         !yIdx.End(); yIdx++) {
+      
+      for (int j = 0; j < axis; j++) {
+        xIdx.Ref(j) = yIdx[j];
       }
-      y_data[i] = x_data[xi];
-//>>>>>>> wlt
+      for (int j = 0; j < indices_ndim; j++) {
+        idxIdx.Ref(j) = yIdx[j + axis];
+      }
+      int32_t axisIndex = indices_data[idxIdx.Index()];
+      xIdx.Ref(axis) = std::min(std::max(axisIndex, 0), (int32_t)X_shape[axis] - 1);
+      for (int j = axis + 1; j < xndim; j++) {
+        xIdx.Ref(j) = yIdx[j + indices_ndim - 1];
+      }
+      y_data[yIdx.Index()] = x_data[xIdx.Index()];
+      xIdx.reset();
+      idxIdx.reset();
```

Comment:
Add formalization comments and it's no need to reset the `xIdx` and `idxIdx`
since you will set all the index values for these variables in each loop.